src/cms/utils/pdf_utils.py

```diff
@@ -0,0 +1,142 @@
+import hashlib
+import logging
+
+from xhtml2pdf import pisa
+
+from django.contrib.staticfiles import finders
+from django.core.cache import cache
+from django.db.models import Min
+from django.http import HttpResponse
+from django.template.loader import get_template
+from django.utils.translation import ugettext as _
+
+from backend.settings import STATIC_URL, MEDIA_URL
+from ..constants import text_directions
+from ..models import Language
+
+logger = logging.getLogger(__name__)
+
+
+# pylint: disable=too-many-locals
+def generate_pdf(region_slug, language_code, region, pages):
+    """
+    Function for handling a pdf export request for pages.
+    The pages were either selected by cms user (see :func:`cms.static.js.pages.page_bulk_action.bulk_action_execute`)
+    or by API request (see :func:`api.v3.pdf_export`)
+    For more information on xhtml2pdf, see :doc:`xhtml2pdf:index`
+
+    :param region_slug: Request submitted for rendering pdf document
+    :type region_slug: ~django.http.HttpRequest
+
+    :param language_code: bcp47 code of the current language
+    :type language_code: str
+
+    :raises ~django.core.exceptions.PermissionDenied: User login and permissions required
+
+    :return: PDF document offered for download
+    :rtype: ~django.http.HttpResponse
+    """
+    # first all necessary data for hashing are collected, starting at region_slug
+    pdf_key_list = [region_slug]
+    for page in pages:
+        # add translation id and last_updated to hash key list if they exist
+        page_translation = page.get_public_translation(language_code)
+        if page_translation:
+            # if translation for this language exists
+            pdf_key_list.append(page_translation.id)
+            pdf_key_list.append(page_translation.last_updated)
+        else:
+            # if the page has no translation for this language
+            pages = pages.exclude(id=page.id)
+    # finally combine all list entries to a single hash key
+    pdf_key_string = "_".join(map(str, pdf_key_list))
+    # compute the hash value based on the hash key
+    pdf_hash = hashlib.sha256(bytes(pdf_key_string, "utf-8")).hexdigest()
+    cached_response = cache.get(pdf_hash, "has_expired")
+    if cached_response != "has_expired":
+        # if django cache already contains a response object
+        return cached_response
+
+    amount_pages = pages.count()
+    language = Language.objects.get(code=language_code)
+    context = {
+        "html2pdf": True,
+        "right_to_left": language.text_direction == text_directions.RIGHT_TO_LEFT,
+        "region": region,
+        "pages": pages,
+        "language": language,
+        "amount_pages": amount_pages,
+    }
+    response = HttpResponse(content_type="application/pdf")
+    if amount_pages == 0:
+        return HttpResponse(_("No valid pages selected for PDF generation."))
```

Comment:
```suggestion
        return HttpResponse(_("No valid pages selected for PDF generation."), status=400)
```
You could also use `400 Bad Request` here (which makes it easier for the client to detect the error.

Also, you could move this condition right after `amount_pages = pages.count()` to make sure no unnecessary code is executed if no pages are given.